addon/PreviewProvider.js

```diff
@@ -208,38 +207,86 @@ PreviewProvider.prototype = {
     let dbLinks = yield this._asyncFindItemsInDB(processedLinks);
     let existingLinks = new Map();
     dbLinks.forEach(item => existingLinks.set(item.cache_key, item));
-    let results = processedLinks.map(site => {
-      if (!site) {
-        return null;
+    let results = [];
+    for (let link of processedLinks) {
+      if (!link) {
+        break;
       }
-      let enhancedLink = Object.assign({}, site);
+      // copy over fields we need from the original site object
+      let enhancedLink = {};
+      enhancedLink.title = link.title;
+      enhancedLink.type = link.type;
+      enhancedLink.url = link.url;
+      enhancedLink.cache_key = link.cache_key;
+      enhancedLink.lastVisitDate = link.lastVisitDate;
+      enhancedLink.bookmarkDateCreated = link.bookmarkDateCreated;
+
+      // get favicon and background color from firefox
+      const firefoxBackgroundColor = yield this._getFaviconColors(link);
+      const firefoxFavicon = link.favicon;
+
+      // get favicon and background color from tippytop
+      const {favicon_url, background_color, metadata_source} = this._tippyTopProvider.processSite(link);
+      const tippyTopFavicon = favicon_url;
+      const tippyTopBackgroundColor = background_color;
```

Comment:
Not sure of the exact Firefox support for this, but I think it should be possible to assign destructured values to a new variable name directly, instead of having to reassign here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names

```js
const {favicon_url: tippyTopFavicon, background_color: tippyTopBackgroundColor, metadata_source} = this._tippyTopProvider.processSite(link);
```